lib/kube/proxy/sess.go

```diff
@@ -0,0 +1,1193 @@
+					User:         s.ctx.User.GetName(),
+					Login:        s.ctx.User.GetName(),
+					Impersonator: s.ctx.Identity.GetIdentity().Impersonator,
+				},
+				ConnectionMetadata: apievents.ConnectionMetadata{
+					RemoteAddr: s.req.RemoteAddr,
+					LocalAddr:  s.sess.kubeAddress,
+					Protocol:   events.EventProtocolKube,
+				},
+				Interactive:               true,
+				Participants:              s.allParticipants(),
+				StartTime:                 sessionStart,
+				EndTime:                   s.forwarder.cfg.Clock.Now().UTC(),
+				KubernetesClusterMetadata: s.ctx.eventClusterMeta(),
+				KubernetesPodMetadata:     eventPodMeta,
+				InitialCommand:            request.cmd,
+				SessionRecording:          s.ctx.recordingConfig.GetMode(),
+			}
+
+			if err := s.emitter.EmitAuditEvent(s.forwarder.ctx, sessionEndEvent); err != nil {
+				s.forwarder.log.WithError(err).Warn("Failed to emit session end event.")
+			}
+		} else {
+			// send an exec event
+			execEvent := &apievents.Exec{
+				Metadata: apievents.Metadata{
+					Type:        events.ExecEvent,
+					ClusterName: s.forwarder.cfg.ClusterName,
+				},
+				ServerMetadata: apievents.ServerMetadata{
+					ServerID:        s.forwarder.cfg.ServerID,
+					ServerNamespace: s.forwarder.cfg.Namespace,
+				},
+				SessionMetadata: apievents.SessionMetadata{
+					SessionID: s.id.String(),
+					WithMFA:   s.ctx.Identity.GetIdentity().MFAVerified,
+				},
+				UserMetadata: apievents.UserMetadata{
+					User:         s.ctx.User.GetName(),
+					Login:        s.ctx.User.GetName(),
+					Impersonator: s.ctx.Identity.GetIdentity().Impersonator,
+				},
+				ConnectionMetadata: apievents.ConnectionMetadata{
+					RemoteAddr: s.req.RemoteAddr,
+					LocalAddr:  s.sess.kubeAddress,
+					Protocol:   events.EventProtocolKube,
+				},
+				CommandMetadata: apievents.CommandMetadata{
+					Command: strings.Join(request.cmd, " "),
+				},
+				KubernetesClusterMetadata: s.ctx.eventClusterMeta(),
+				KubernetesPodMetadata:     eventPodMeta,
+			}
+
+			if err != nil {
+				execEvent.Code = events.ExecFailureCode
+				execEvent.Error = err.Error()
+				if exitErr, ok := err.(utilexec.ExitError); ok && exitErr.Exited() {
+					execEvent.ExitCode = fmt.Sprintf("%d", exitErr.ExitStatus())
+				}
+			} else {
+				execEvent.Code = events.ExecCode
+			}
+
+			if err := s.emitter.EmitAuditEvent(s.forwarder.ctx, execEvent); err != nil {
+				s.forwarder.log.WithError(err).Warn("Failed to emit event.")
+			}
+		}
+	}()
+
+	options := remotecommand.StreamOptions{
+		Stdin:             s.io,
+		Stdout:            s.io,
+		Stderr:            s.io,
+		Tty:               request.tty,
+		TerminalSizeQueue: s.terminalSizeQueue,
+	}
+
+	if err = executor.Stream(options); err != nil {
+		s.log.WithError(err).Warning("Executor failed while streaming.")
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// join attempts to connect a party to the session.
+func (s *session) join(p *party) error {
+	s.mu.Lock()
```

Comment:
Same here - there's quite a lot happening while we hold the mutex. Is it necessary to hold it the whole time or can some of this logic be moved outside the critical section?